deployer/resource_allocation/update_nodeinfo.py

```diff
@@ -0,0 +1,146 @@
+    # container runtime, systemd, etc. This does *not* count for daemonsets
+    # and pods runninng on the kubernetes cluster.
+    allocatable = node["status"]["allocatable"]
+    cpu_allocatable = parse_quantity(allocatable["cpu"])
+    mem_allocatable = parse_quantity(allocatable["memory"])
+
+    # Find all pods running on this node
+    all_pods = json.loads(
+        subprocess.check_output(
+            [
+                "kubectl",
+                "get",
+                "pod",
+                "-A",
+                "--field-selector",
+                f'spec.nodeName={node["metadata"]["name"]}',
+                "-o",
+                "json",
+            ]
+        ).decode()
+    )["items"]
+
+    # Filter out jupyterhub user pods
+    # TODO: Filter out dask scheduler and worker pods
+    pods = [
+        p
+        for p in all_pods
+        if p["metadata"]["labels"].get("component") not in ("singleuser-server",)
+    ]
+
+    # This is the amount of resources available for our workloads - jupyter and dask.
+    # We start with the allocatable resources, and subtract the resource *requirements*
+    # for all the pods that are running on every node, primarily from kube-system and
+    # support. The amount left over is what is available for the *scheduler* to put user pods
+    # on to.
+    cpu_available = cpu_allocatable
+    mem_available = mem_allocatable
+
+    for p in pods:
+        mem_request = 0
+        cpu_request = 0
+        # Iterate through all the containers in the pod, and count the memory & cpu requests
+        # they make. We don't count initContainers' requests as they don't overlap with the
+        # container requests at any point.
+        for c in p["spec"]["containers"]:
+            mem_request += parse_quantity(
+                c.get("resources", {}).get("requests", {}).get("memory", "0")
+            )
+            cpu_request += parse_quantity(
+                c.get("resources", {}).get("requests", {}).get("cpu", "0")
+            )
+        cpu_available -= cpu_request
+        mem_available -= mem_request
+
+    # Add a little bit of wiggle room, to account for:
+    # 1. Changes in requests for system components as k8s versions upgrade or
+    #    cloud providers roll out new components
+    # 2. We deploy support components but forget to update node capacity info
+    # 3. Whatever other things we aren't currently thinking of.
+    # A small amount of memory and CPU to sacrifice for the sake of
+    # operational flexibility. However, we *must* regenerate and update node
+    # information each time the following events occur:
+    # 1. We upgrade kubernetes versions
+    # 2. We change resource requirements for *daemonsets* in our support chart.
+    # 3. We upgrade z2jh version
+
+    # 128 MiB memory buffer
+    mem_available = mem_available - (128 * 1024 * 1024)
+    # 0.05 CPU overhead
+    cpu_available = cpu_available - 0.05
```

Comment:
I'm leaning heavily towards ensuring that we don't run into "fail to start server" issues caused by this over minimizing the headroom. Can we make it at least something like 256 Mi and 0.15 CPU headroom?

I think a memory overhead of 128 is cutting it closer than merited given our nodes have 32GB+ and a failure in getting this right can cause an issue that we won't observe before we have a runtime failure detected by users. By not cutting it close, the memory can also help provide a buffer for when the node includes workloads that don't have requests=limits, so its also not wasted.

I think the cpu overhead is too small as well, mostly because I don't see a good enough motivation of cutting it close since we the CPU overhead won't go wasted as long as we don't put super tight CPU limits without.

Reply:
Thank you for writing this out clearly, @consideRatio.

> I think the cpu overhead is too small as well, mostly because I don't see a good enough motivation of cutting it close since we the CPU overhead won't go wasted as long as we don't put super tight CPU limits without.

I think this makes 100% sense for the one strategy introduced in this PR, as it only tackles cpu requests not limits. So I'll increase the headroom to 0.15 (or 0.20) but move the headroom calculation to the strategy code, when we introduce other strategies in the future, they can make their own choices.

I'm trying to understand *when* exactly this will actually cause a "fail to start server" issue, rather than an issue where utilization of a server is not 100% when packed. I think this only happens at the case where the resource allocation sets guarantee to 100% of available memory. Looking at the set of choices currently deployed to openscapes, that's 2 of the 6 choices. So I understand in those cases, if circumstances change in such a way that the *static* calculation here for 'available' doesn't match, users would end up with a hanging pod that never gets scheduled anywhere. In the other 4 choices, what will happen instead is wastage of resources, as a new node will be spun up when a user might have already fit in the previous node.

So to handle the memory *requests* case, I will do the following:

1. Increase the general overhead calculation, from which *all* resource allocations are measured, from 128Mi to 256Mi. 
2. Specifically for the choices where we are allocating a full node, increase this number *even further* (probably to 512Mi), as a measurement mismatch here would cause a bigger user problem than with (1).
3. Open an issue to consider changing the strategy to have *limits* that don't count the overhead, and *requests* that do count the overhead. However, this makes the process a little more complex, and so I'd prefer to not do this on the first run. Sacrificing a little bit of extra RAM for the simplicity seems a worthwhile tradeoff.

I think over time, as we gain more experience with this strategy as well as data, we can fine tune this some more. Ideally, the "mem_available" data would be *dynamic*, rather than *static* - I think this would increase confidence level that this drift would not occur. I'll think of ideas how this can be done, but won't block progress on this PR on that.

Does this satisfy you, @consideRatio?

Reply:
Thank you for that chat Yuvi!! I opened #3132 to describe my proposed simple strategy